README.md
# Clockwork

Clockwork is a Go library which allows you to send SMS messages using the [Clockwork SMS service](https://www.clockworksms.com/).

# Installation

Requires Go version 1.7+.

> $ go get github.com/umahmood/clockwork
>
> $ cd $GOPATH/src/github.com/umahmood/clockwork
>
> $ go test ./...

# Usage

Send an SMS message:
```
package main

import (
    "fmt"

    "github.com/umahmood/clockwork"
)

func main() {
    cw := clockwork.New("API-KEY")

    msg := clockwork.SMS{
        To:      clockwork.Numbers{"44123456789", "44987654321"},
        From:    "Gopher",
        Content: "Gophers rule!",
    }

    resp, err := cw.Send(msg)
    if err != nil {
        // ...
    }

    // print each valid number an SMS was sent to and its assigned message id
    for num, meta := range resp {
        fmt.Println(num, meta["ID"])
    }
}
```

Delivery receipts let you know whether a message has been delivered:
```
package main

import (
    "fmt"

    "github.com/umahmood/clockwork"
)

func OnDeliveryReceipt(r clockwork.Receipt) {
    if r.Err != nil {
        //...
    }
    fmt.Println("ID", r.ID, "Status", r.Status, "To", r.To, "Time", r.Time)
}

func main() {
    clockwork.DeliveryReceiptListen(&clockwork.ReceiptHandler{
        Path:     "/receipts",
        Port:     9090,
        Callback: OnDeliveryReceipt,
    })
}
```
Output:
```
ID LA_422342 Status Enroute To 441234567890 Time 2017-03-17 17:21:27.667238018 +0000 UTC
```

# Documentation

> http://godoc.org/github.com/umahmood/clockwork

# License

See the [LICENSE](LICENSE.md) file for license rights and limitations (MIT).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.2.1] - 2017-03-17
### Added
- Time field to delivery receipts

## [1.1.1] - 2017-03-12
### Added
- Receive delivery receipts, let you know whether a message has been delivered

## [1.0.1] - 2017-02-28
### Fixed
- Moved hard coded regex patterns into a separate file


clockwork_test/clockwork_test.go
package clockwork_test

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"
	"testing"
	"time"

	"github.com/umahmood/clockwork"
)

const testAPIKey = "TEST-KEY"

// mockClockwork test instance
type mockClockwork struct {
	f      func(req *http.Request) (*http.Response, error)
	apiKey string
}

// Send mock SMS
func (m *mockClockwork) Send(sms clockwork.SMS) (clockwork.SMSResponse, error) {
	return clockwork.DoSendRequestHelper(m, m.apiKey, "https://test.com/send", sms)
}

// Credit mock credit
func (m *mockClockwork) Credit() (amount float64, code string, err error) {
	return clockwork.DoCreditRequestHelper(m, m.apiKey, "https://test.com/credit")
}

// Do mock HTTP request
func (m *mockClockwork) Do(req *http.Request) (*http.Response, error) {
	return m.f(req)
}

// TestSingleValidSMSNumber
func TestSingleValidSMSNumber(t *testing.T) {
	mock := &mockClockwork{
		apiKey: testAPIKey,
		f: func(req *http.Request) (*http.Response, error) {
			body := "To: 1234567890 ID: VE_439333520"
			return &http.Response{
				StatusCode: 200,
				Body:       ioutil.NopCloser(strings.NewReader(body)),
			}, nil
		},
	}

	msg := clockwork.SMS{
		To:      clockwork.Numbers{"1234567890"},
		From:    "Gopher",
		Content: "Gophers rule!",
	}

	wantNumber := "1234567890"
	wantID := "VE_439333520"

	resp, err := mock.Send(msg)
	if err != nil {
		t.Error(err)
	}

	for gotNumber, gotInfo := range resp {
		if gotNumber != wantNumber {
			t.Errorf("Fail: got %s want %s", gotNumber, wantNumber)
		}
		gotID := gotInfo["ID"]
		if gotID != wantID {
			t.Errorf("Fail: got %s want %s", gotID, wantID)
		}
	}
}

// TestSingleInvalidSMSNumber
func TestSingleInvalidSMSNumber(t *testing.T) {
	mock := &mockClockwork{
		apiKey: testAPIKey,
		f: func(req *http.Request) (*http.Response, error) {
			body := "To: 123 Error 10: Invalid 'To' Parameter"
			return &http.Response{
				StatusCode: 200,
				Body:       ioutil.NopCloser(strings.NewReader(body)),
			}, nil
		},
	}

	msg := clockwork.SMS{
		To:      clockwork.Numbers{"123"},
		From:    "Gopher",
		Content: "Gophers rule!",
	}

	resp, err := mock.Send(msg)
	if err != clockwork.ErrInvalidTo {
		t.Errorf("Fail: err - got %s want %s", err, clockwork.ErrInvalidTo)
	}

	if resp != nil {
		t.Errorf("Fail: resp - got %v want nil", resp)
	}
}

// TestBothValidAndInvalidNumbers
func TestBothValidAndInvalidNumbers(t *testing.T) {
	mock := &mockClockwork{
		apiKey: testAPIKey,
		f: func(req *http.Request) (*http.Response, error) {
			body := `To: 123 Error 10: Invalid 'To' Parameter
					 To: 456 Error 10: Invalid 'To' Parameter
					 To: 13053696625 ID: LA_360224205
					 To: 44123456789 ID: VE_360224242`
			return &http.Response{
				StatusCode: 200,
				Body:       ioutil.NopCloser(strings.NewReader(body)),
			}, nil
		},
	}

	msg := clockwork.SMS{
		To:      clockwork.Numbers{"123", "456", "13053696625", "44123456789"},
		From:    "Gopher",
		Content: "Gophers rule!",
	}

	resp, err := mock.Send(msg)
	if err != clockwork.ErrInvalidTo {
		t.Errorf("Fail: err - got %s want %s", err, clockwork.ErrInvalidTo)
	}

	if resp["123"] != nil {
		t.Errorf("Fail: invalid number 123 in map")
	} else if resp["456"] != nil {
		t.Errorf("Fail: invalid number 456 in map")
	}

	if resp["13053696625"] == nil {
		t.Errorf("Fail: valid number 13053696625 not in map")

	} else if resp["44123456789"] == nil {
		t.Errorf("Fail: valid number 44123456789 not in map")
	}

	if len(resp) != 2 {
		t.Errorf("Fail: len map got %d want 2", len(resp))
	}
}

// TestCredit
func TestCredit(t *testing.T) {
	testCases := []struct {
		body       string
		wantCode   string
		wantCredit float64
	}{
		{
			body:       "Balance: 2048.12 (GBP)",
			wantCode:   "GBP",
			wantCredit: 2048.12,
		},
		{
			body:       "Balance: 234.56 (GBP)",
			wantCode:   "GBP",
			wantCredit: 234.56,
		},
		{
			body:       "Balance: 42.58 (GBP)",
			wantCode:   "GBP",
			wantCredit: 42.58,
		},
		{
			body:       "Balance: 1.42 (GBP)",
			wantCode:   "GBP",
			wantCredit: 1.42,
		},
		{
			body:       "Balance: 0.45 (GBP)",
			wantCode:   "GBP",
			wantCredit: 0.45,
		},
		{
			body:       "Balance: 0.00 (GBP)",
			wantCode:   "GBP",
			wantCredit: 0.00,
		},
		{
			body:       "Balance: 0 (GBP)",
			wantCode:   "GBP",
			wantCredit: 0,
		},
		{
			body:       "Balance: -94.23 (GBP)",
			wantCode:   "GBP",
			wantCredit: -94.23,
		},
	}

	for _, tc := range testCases {
		t.Run(fmt.Sprintf("input_%s", tc.body), func(t *testing.T) {
			mock := &mockClockwork{
				apiKey: testAPIKey,
				f: func(req *http.Request) (*http.Response, error) {
					body := tc.body
					return &http.Response{
						StatusCode: 200,
						Body:       ioutil.NopCloser(strings.NewReader(body)),
					}, nil
				},
			}
			gotCredit, gotCode, err := mock.Credit()
			if err != nil {
				t.Errorf("Fail err - got %v want nil", err)
			}

			if gotCredit != tc.wantCredit {
				t.Errorf("Fail credit - got %v want %v", gotCredit, tc.wantCredit)
			}

			if gotCode != tc.wantCode {
				t.Errorf("Fail code - got %s want %s", gotCode, tc.wantCode)
			}

		})
	}
}

// TestCreditError
func TestCreditError(t *testing.T) {
	mock := &mockClockwork{
		apiKey: testAPIKey,
		f: func(req *http.Request) (*http.Response, error) {
			body := "Error 58: Invalid API Key"
			return &http.Response{
				StatusCode: 200,
				Body:       ioutil.NopCloser(strings.NewReader(body)),
			}, nil
		},
	}

	wantError := clockwork.ErrInvalidAPIKey

	gotCredit, gotCode, gotError := mock.Credit()

	if gotCredit != 0 {
		t.Errorf("Fail: credit got %v want %v", gotCredit, 0)
	}

	if gotCode != "" {
		t.Errorf("Fail: code got %s want %s", gotCode, "")
	}

	if gotError != wantError {
		t.Errorf("Fail: error got %v want %v", gotError, wantError)
	}
}

// TestDeliveryReceipts
func TestDeliveryReceipts(t *testing.T) {
	onDeliveryReceipt := func(got clockwork.Receipt) {
		var wantID = "LA_424242"
		var wantTo = "441234567890"
		var wantStatus = clockwork.Delivered
		var wantTime = time.Now().UTC()
		var wantErr error

		if got.ID != wantID {
			t.Errorf("Fail: id - got %s want %s", got.ID, wantID)
		}

		if got.To != wantTo {
			t.Errorf("Fail: to - got %s want %s", got.To, wantTo)
		}

		if got.Status != wantStatus {
			t.Errorf("Fail: status - got %v want %v", got.Status, wantStatus)
		}

		if got.Time.Year() != wantTime.Year() || got.Time.Month() != wantTime.Month() || got.Time.Day() != wantTime.Day() {
			t.Errorf("Fail: time - got %v want %v", got.Time, wantTime)
		}

		if got.Err != wantErr {
			t.Errorf("Fail: err - got %v want %v", got.Err, wantErr)
		}
	}

	testPath := "/receipts"
	testPort := 9090
	testQueryParams := "msg_id=LA_424242&status=DELIVRD&detail=0&to=441234567890"

	testURL := fmt.Sprintf("http://localhost:%d%s?%s", testPort, testPath, testQueryParams)

	go func() {
		clockwork.DeliveryReceiptListen(&clockwork.ReceiptHandler{
			Path:     testPath,
			Port:     testPort,
			Callback: onDeliveryReceipt,
		})
	}()

	testCases := []struct {
		method string
		url    string
		buf    *bytes.Buffer
	}{
		{
			method: "GET",
			url:    testURL,
			buf:    bytes.NewBufferString(""),
		},
		{
			method: "POST",
			url:    testURL,
			buf:    bytes.NewBufferString(testQueryParams),
		},
	}

	for _, tc := range testCases {
		req, err := http.NewRequest(tc.method, tc.url, tc.buf)
		if err != nil {
			t.Errorf("Fail: err %v", err)
		}
		client := &http.Client{}
		client.Do(req)
	}
}


errors.go
package clockwork

import "errors"

// Errors related to delivery receipts. // Check the following link for more
// information:
// https://www.clockworksms.com/doc/reference/faqs/delivery-states/

var (
	// ErrNoNetworkDetails no details provided by network
	ErrNoNetworkDetails = errors.New("clockwork: no details provided by network")

	// ErrMessageDetailsWrong message details wrong
	ErrMessageDetailsWrong = errors.New("clockwork: message details wrong")

	// ErrPermOperator permanent operator error
	ErrPermOperator = errors.New("clockwork: operator error - permanent")

	// ErrTempOperator temporary operator error
	ErrTempOperator = errors.New("clockwork: operator error - temporary")

	// ErrPermAbsentSub permanent agent subscriber error
	ErrPermAbsentSub = errors.New("clockwork: absent subscriber - permanent")

	// ErrTempAbsentSub temporary agent subscriber error
	ErrTempAbsentSub = errors.New("clockwork: absent subscriber - temporary")

	// ErrPermPhone permanent phone related error
	ErrPermPhone = errors.New("clockwork: phone related error - permanent")

	// ErrTempPhone temporary phone related error
	ErrTempPhone = errors.New("clockwork: phone related error - temporary")
)

// You might get one of the following errors back when making a call to the API.
// Check the following link for more information:
// https://www.clockworksms.com/doc/reference/faqs/api-error-codes/

var (
	// ErrInternal internal error something went wrong in our API
	ErrInternal = errors.New("clockwork: internal error")

	// ErrInvalidUsernamePassword invalid username or password
	ErrInvalidUsernamePassword = errors.New("clockwork: invalid username or password")

	// ErrInsufficientCredit insufficient credits available
	ErrInsufficientCredit = errors.New("clockwork: insufficient credits available")

	// ErrAuthFail authentication failure
	ErrAuthFail = errors.New("clockwork: authentication failure")

	// ErrInvalidMsgType invalid 'MsgType' parameter
	ErrInvalidMsgType = errors.New("clockwork: invalid 'MsgType'")

	// ErrMissingTo 'To' parameter not specified
	ErrMissingTo = errors.New("clockwork: 'To' parameter not specified")

	// ErrMissingContent 'Content' parameter not specified
	ErrMissingContent = errors.New("clockwork: 'Content' parameter not specified")

	// ErrMissingMessageID 'MessageID' parameter not specified
	ErrMissingMessageID = errors.New("clockwork: 'MessageID' parameter not specified")

	// ErrUnknownMessageID unknown 'MessageID'
	ErrUnknownMessageID = errors.New("clockwork: unknown 'MessageID'")

	// ErrInvalidTo invalid 'To' parameter
	ErrInvalidTo = errors.New("clockwork: invalid 'To' parameter some or all numbers were not correct")

	// ErrInvalidFrom invalid 'From' parameter
	ErrInvalidFrom = errors.New("clockwork: invalid 'From' parameter")

	// ErrMessageTooLong message text is too long
	ErrMessageTooLong = errors.New("clockwork: message text is too long")

	// ErrRoutingMessage cannot route message
	ErrRoutingMessage = errors.New("clockwork: cannot route message")

	// ErrMessageExpired message expired
	ErrMessageExpired = errors.New("clockwork: message expired")

	// ErrNoRoute no route defined for this number
	ErrNoRoute = errors.New("clockwork: no route defined for this number")

	// ErrMissingURL 'URL' parameter not set
	ErrMissingURL = errors.New("clockwork: 'URL' parameter not set")

	// ErrInvalidSourceIP invalid source IP
	ErrInvalidSourceIP = errors.New("clockwork: invalid source IP")

	// ErrMissingUDH 'UDH' parameter not specified
	ErrMissingUDH = errors.New("clockwork: 'UDH' parameter not specified")

	// ErrInvalidServType invalid 'ServType' parameter
	ErrInvalidServType = errors.New("clockwork: invalid 'ServType' parameter")

	// ErrInvalidExpiryTime invalid 'ExpiryTime' parameter
	ErrInvalidExpiryTime = errors.New("clockwork: invalid 'ExpiryTime' parameter")

	// ErrDuplicateClientID duplicate 'ClientID' received
	ErrDuplicateClientID = errors.New("clockwork: duplicate 'ClientID' received")

	// ErrInvalidTimeStamp invalid ‘TimeStamp’ parameter
	ErrInvalidTimeStamp = errors.New("clockwork: invalid ‘TimeStamp’ parameter")

	// ErrInvalidAbsExpiry invalid ‘AbsExpiry’ parameter
	ErrInvalidAbsExpiry = errors.New("clockwork: invalid ‘AbsExpiry’ parameter")

	// ErrInvalidDlrType invalid 'DlrType' parameter
	ErrInvalidDlrType = errors.New("clockwork: invalid 'DlrType' parameter")

	// ErrInvalidConcat invalid 'Concat' parameter
	ErrInvalidConcat = errors.New("clockwork: invalid 'Concat' parameter")

	// ErrInvalidUniqueID invalid 'UniqueID' parameter
	ErrInvalidUniqueID = errors.New("clockwork: invalid 'UniqueId' parameter")

	// ErrClientIDRequired client id required. your account is setup to check for a
	// unique client id on every message, one wasn't supplied in this send.
	ErrClientIDRequired = errors.New("clockwork: 'ClientID' Required - your account is setup to check for a unique client ID on every message, one wasn't supplied in this send")

	// ErrInvalidCharInContent invalid character in 'Content' parameter
	ErrInvalidCharInContent = errors.New("clockwork: invalid character in 'Content' parameter")

	// ErrInvalidTextPayload invalid 'Text' Payload MMS text has an invalid character
	ErrInvalidTextPayload = errors.New("clockwork: invalid 'TextPayload' MMS text has an invalid character")

	// ErrInvalidHexPayload invalid 'Hex' Payload MMS payload can't be decoded as hex
	ErrInvalidHexPayload = errors.New("clockwork: invalid 'HexPayload' MMS Payload cant be decoded as hex")

	// ErrInvalidBase64Payload invalid 'Base64' Payload MMS payload can't be decoded
	// as base64.
	ErrInvalidBase64Payload = errors.New("clockwork: invalid 'Base64Payload' MMS payload cant be decoded as base64")

	// ErrMissingContentType missing content type. No content type provided on MMS
	// payload.
	ErrMissingContentType = errors.New("clockwork: missing 'ContentType'")

	// ErrMissingID missing 'ID' all MMS payload parts must have an id
	ErrMissingID = errors.New("clockwork: missing 'ID' all MMS Payload parts must have an ID")

	// ErrMMSMessageTooLarge MMS message too large the combined parts are too large
	// to send
	ErrMMSMessageTooLarge = errors.New("clockwork: The combined parts are too large to send")

	// ErrInvalidPayloadID invalid 'Payload' id
	ErrInvalidPayloadID = errors.New("clockwork: invalid payload ID")

	// ErrDuplicatePayloadID duplicate payload id
	ErrDuplicatePayloadID = errors.New("clockwork: duplicate payload ID")

	// ErrNoPayloadOnMMS no payload on MMS
	ErrNoPayloadOnMMS = errors.New("clockwork: no payload on MMS")

	// ErrDuplicateFileName duplicate 'filename' attribute on payload. All MMS parts
	// must have unique filenames.
	ErrDuplicateFileName = errors.New("clockwork: duplicate 'filename' attribute on payload -all MMS parts must have unique filenames")

	// ErrMissingItemID 'ItemId' parameter not specified
	ErrMissingItemID = errors.New("clockwork: 'ItemId' parameter not specified")

	// ErrInvalidItemID invalid 'ItemId' parameter
	ErrInvalidItemID = errors.New("clockwork: invalid 'ItemId' parameter")

	// ErrGenerateFileName unable to generate filename for content-type
	ErrGenerateFileName = errors.New("clockwork: unable to generate filename for 'Content-Type'")

	// ErrInvalidCharAction invalid 'InvalidCharAction' parameter
	ErrInvalidCharAction = errors.New("clockwork: invalid 'InvalidCharAction' parameter")

	// ErrInvalidDlrEnroute invalid 'DlrEnroute' parameter
	ErrInvalidDlrEnroute = errors.New("clockwork: invalid 'DlrEnroute' parameter")

	// ErrInvalidTruncate invalid 'Truncate' parameter
	ErrInvalidTruncate = errors.New("clockwork: invalid 'Truncate' parameter")

	// ErrInvalidLong invalid 'Long' parameter
	ErrInvalidLong = errors.New("clockwork: invalid 'Long' parameter")

	// ErrNoAPIKey no API key provided. You need to provide an API key or a
	// username and password when calling the API.
	ErrNoAPIKey = errors.New("clockwork: no API key provided you need to provide an API Key or a user name and password when calling the API")

	// ErrInvalidAPIKey invalid API key. Log in to your API account to check the key
	// or create a new one
	ErrInvalidAPIKey = errors.New("clockwork: invalid API key - log in to your API account to check the key or create a new one")

	// ErrMustUseAPIKeys account must use API keys. This account isn't allowed to
	// use a username and password, log in to your account to create a key.
	ErrMustUseAPIKeys = errors.New("clockwork: account must use API keys")

	// ErrBlockedSpam blocked by Spam filter. Sometimes your message will be caught
	// by our Spam filter. If you're having trouble because of this error - get in
	// touch.
	ErrBlockedSpam = errors.New("clockwork: blocked by spam filter - sometimes messages will be caught by our spam filter")

	// ErrInvalidXML invalid XML API post can't be parsed as XML.
	ErrInvalidXML = errors.New("clockwork: invalid XML API post can't be parsed as XML")

	// ErrInvalidXMLDoc XML document does not validate
	ErrInvalidXMLDoc = errors.New("clockwork: XML document does not validate")

	// ErrLongClientID client id too long
	ErrLongClientID = errors.New("clockwork: Client ID too long")

	// ErrRateExceeded query throttling rate exceeded. You've sent too many status
	// requests this hour
	ErrRateExceeded = errors.New("clockwork: query throttling rate exceeded - you've sent too many status requests this hour")

	// ErrStatusCode Clockwork SMS API returned a non 200 HTTP status code
	ErrStatusCode = errors.New("clockwork: API request did not return HTTP 200 OK")

	// ErrUnknown if error is not in errorMap then this error will be returned
	ErrUnknown = errors.New("clockwork: unknown API error code")
)

// errorMap maps Clockwork API error codes to error messages. The keys (numbers)
// are important, they match the API error codes documented here:
// https://www.clockworksms.com/doc/reference/faqs/api-error-codes/
var errorMap = map[int]error{
	1:  ErrInternal,
	2:  ErrInvalidUsernamePassword,
	3:  ErrInsufficientCredit,
	4:  ErrAuthFail,
	5:  ErrInvalidMsgType,
	6:  ErrMissingTo,
	7:  ErrMissingContent,
	8:  ErrMissingMessageID,
	9:  ErrUnknownMessageID,
	10: ErrInvalidTo,

	11: ErrInvalidFrom,
	12: ErrMessageTooLong,
	13: ErrRoutingMessage,
	14: ErrMessageExpired,
	15: ErrNoRoute,
	16: ErrMissingURL,
	17: ErrInvalidSourceIP,
	18: ErrMissingUDH,
	19: ErrInvalidServType,
	20: ErrInvalidExpiryTime,

	25: ErrDuplicateClientID,
	26: ErrInternal,
	27: ErrInvalidTimeStamp,
	28: ErrInvalidAbsExpiry,
	29: ErrInvalidDlrType,

	31: ErrInvalidConcat,
	32: ErrInvalidUniqueID,
	33: ErrClientIDRequired,

	39: ErrInvalidCharInContent,
	40: ErrInvalidTextPayload,
	41: ErrInvalidHexPayload,
	42: ErrInvalidBase64Payload,
	43: ErrMissingContentType,
	44: ErrMissingID,
	45: ErrMMSMessageTooLarge,
	46: ErrInvalidPayloadID,
	47: ErrDuplicatePayloadID,
	48: ErrNoPayloadOnMMS,
	49: ErrDuplicateFileName,
	50: ErrMissingItemID,
	51: ErrInvalidItemID,
	52: ErrGenerateFileName,
	53: ErrInvalidCharAction,
	54: ErrInvalidDlrEnroute,
	55: ErrInvalidTruncate,
	56: ErrInvalidLong,
	57: ErrNoAPIKey,
	58: ErrInvalidAPIKey,
	59: ErrMustUseAPIKeys,
	60: ErrBlockedSpam,

	100: ErrInternal,
	101: ErrInternal,
	102: ErrInvalidXML,
	103: ErrInvalidXMLDoc,

	300: ErrLongClientID,
	305: ErrRateExceeded,
}

// errorFromCode looks up and returns the correct Clockwork API error. If the
// error code is unknown, returns ErrUnknown.
func errorFromCode(c int) error {
	if err, ok := errorMap[c]; ok {
		if err != nil {
			return err
		}
	}
	return ErrUnknown
}


clockwork.go
package clockwork

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// SendURL Clockwork SMS send http end point
const SendURL = "https://api.clockworksms.com/http/send.aspx"

// CreditURL Clockwork SMS credit http end point
const CreditURL = "https://api.clockworksms.com/http/balance"

// MsgType options
const (
	// TEXT Standard SMS message. Any character from the GSM character set can
	// be used. A single SMS can contain 160 standard GSM characters.
	// https://www.clockworksms.com/doc/reference/faqs/gsm-character-set/
	TEXT = "TEXT"
	// UCS2 Unicode SMS Message. Any Unicode characters from the UCS-2 character
	// set can be sent.
	UCS2 = "UCS2"
)

// Concat options
const (
	// OnePart SMS can contain 160 standard GSM characters
	OnePart = iota + 1
	// TwoParts SMS can contain 306 standard GSM characters
	TwoParts
	// ThreeParts SMS can contain 459 standard GSM characters
	ThreeParts
)

// InvalidCharAction options
const (
	// ErrorOnInvalidChars if invalid characters in the message are encountered
	ErrorOnInvalidChars = iota + 1
	// RemoveInvalidChars invalid characters in the message if encountered
	RemoveInvalidChars
	// ReplaceInvalidChars invalid characters where possible, remove the rest
	ReplaceInvalidChars
)

// Truncate options
const (
	// ErrorIfContentTooLong return error if content is too long
	ErrorIfContentTooLong = iota + 1
	// ReplaceExtraText remove the extra text
	ReplaceExtraText
)

// SMS represents a single SMS message
type SMS struct {
	// To list of up to 50 numbers. Phone numbers must be in international number
	// format without a leading ‘+’ or international dialing prefix such as ‘00’,
	// e.g. 441234567890.
	To []string
	// Content the message you want to send. Mobile networks only support
	// characters listed in the GSM character set.
	Content string
	// The text or phone number displayed when a text message is received on a
	// phone. This can be either a 12 digit number or 11 characters long. You
	// can set a default by logging in to Clockwork.
	From string
	// MsgType message type the default is TEXT.
	MsgType string
	// Concat The maximum number of parts for concatenated messages. Defaults to
	// 1 part, maximum 3. This parameter only affects TEXT message types. Each
	// part is billed as an individual message. When set to 1 the platform will
	// only allow 1 SMS to be sent. When set to 2, it allows either one or two
	// SMS to be sent. When set to 3, the platform allows 1,2 or 3 SMS to be
	// sent depending on the size of the content. Possible values - OnePart,
	// TwoParts, ThreeParts.
	Concat int
	// ClientID a unique Message ID specified by the connecting application,
	// maximum length: 50 characters.
	ClientID string
	// Expiry the number of minutes before the message expires. The minimum is
	// 10 and the maximum is 2160 (36 hours). Expiry time may not be honored by
	// some mobile networks.
	Expiry time.Duration
	// AbsExpiry the Absolute Expiry time for the message. An absolute time
	// should be specified in UTC. Expiry time may not be honored by some mobile
	// networks.
	AbsExpiry time.Time
	// UniqueIDChecks enable unique ID checks. If enabled, the ClientID specified
	// by the connecting application must be unique within the last 12 hours.
	// This is to prevent the connecting application from falsely sending
	// duplicate messages to a phone.
	UniqueIDChecks bool
	// InvalidCharAction what to do with any invalid characters in the message
	// content. Possible values - Error, Remove, Replace.
	InvalidCharAction int
	// Truncate trims the message content to the maximum length if it’s too
	// long. Truncate only works with standard text messages (MsgType=TEXT).
	// Possible values - ErrorIfContentTooLong, ReplaceExtraText.
	Truncate int
}

// SMSResponse stores telephone numbers and there associated meta data
type SMSResponse map[string]map[string]string

// Doer implemented by any type which can do HTTP requests
type Doer interface {
	Do(*http.Request) (*http.Response, error)
}

// Numbers slice of telephone numbers
type Numbers []string

// Clockwork instance
type Clockwork struct {
	apiKey string
}

// New creates a new instance of Clockwork SMS
func New(apiKey string) *Clockwork {
	return &Clockwork{
		apiKey: apiKey,
	}
}

// Send sends a single SMS message. If a message contains a mixture of valid and
// invalid numbers, then the method will return ErrInvalidTo. The valid numbers
// will be in the 'SMSResponse' map. To determine which numbers are invalid,
// check for there existent in the 'SMSResponse' map. For example:
//
//      msg := clockwork.SMS{
//              To: clockwork.Numbers{  "13052645330",   // valid
//                                      "4412345678910", // valid
//                                      "123",           // invalid
//                                      "meh!",          // invalid :)
//                                  }
//              ...
//              }
//
//      resp, err := cw.Send(msg)
//      if err == ErrInvalidTo {
//          // all/some of the numbers are invalid
//			// ...
//      }
//
//      // 'resp' map holds only the valid numbers SMS messages were sent to
//      resp["13052645330"] == true
//      resp["meh!"] == false
//
func (c *Clockwork) Send(sms SMS) (SMSResponse, error) {
	return DoSendRequestHelper(c, c.apiKey, SendURL, sms)
}

// Credit check how much credit you have left on your account
func (c *Clockwork) Credit() (credit float64, code string, err error) {
	return DoCreditRequestHelper(c, c.apiKey, CreditURL)
}

// Do performs a HTTP request
func (c *Clockwork) Do(req *http.Request) (*http.Response, error) {
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

// DoSendRequestHelper helper to make a HTTP Get request using 'sms' values
func DoSendRequestHelper(d Doer, key string, url string, sms SMS) (SMSResponse, error) {
	m := smsSetOptions(sms)
	m["Key"] = key
	q := urlEncode(m)

	req, err := http.NewRequest("GET", url+"?"+q, nil)
	if err != nil {
		return nil, err
	}

	// identify this client to the clockwork SMS API
	userAgent := "Clockwork Go wrapper/" + Version()
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("Content-Type", "text/plain; charset=utf-8")

	resp, err := d.Do(req)
	if err != nil {
		return nil, err
	}

	if resp != nil {
		defer resp.Body.Close()
	}

	if resp.StatusCode != 200 {
		return nil, ErrStatusCode
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return parseSendResponseBody(string(body))
}

// DoCreditRequestHelper helper to make a HTTP Get request
func DoCreditRequestHelper(d Doer, key string, url string) (credit float64, code string, err error) {

	req, err := http.NewRequest("GET", url+"?key="+key, nil)
	if err != nil {
		return 0, "", err
	}

	// identify this client to the clockwork SMS API
	userAgent := "Clockwork Go wrapper/" + Version()
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("Content-Type", "text/plain; charset=utf-8")

	resp, err := d.Do(req)
	if err != nil {
		return 0, "", err
	}

	if resp != nil {
		defer resp.Body.Close()
	}

	if resp.StatusCode != 200 {
		return 0, "", ErrStatusCode
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return 0, "", err
	}

	return parseCreditResponseBody(string(body))
}

// parseSendResponseBody parse the plain text response body from a clockwork
// /send HTTP call.
func parseSendResponseBody(body string) (SMSResponse, error) {
	// if some of the numbers provided to SMS message contained bad numbers,
	// flag it. So we can return the correct error to the caller.
	var badNumbers bool
	nums := make(SMSResponse)
	scanner := bufio.NewScanner(strings.NewReader(body))
	scanner.Split(bufio.ScanLines)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "Error") {
			// error response
			if strings.Contains(line, "To:") {
				// special case - if user specified invalid 'To' number, error
				// response is formatted as:
				//
				//       To: 123 Error 10: Invalid 'To' Parameter
				//       To: 456 Error 10: Invalid 'To' Parameter
				//       To: 441234567890 ID: VE_439221450
				//
				// the first two lines are bad numbers the last line is a valid
				// number.
				badNumbers = true
				continue
			} else {
				// all other error responses are formated as:
				//
				//      Error <number>: <message>
				//
				r := regexp.MustCompile(matchErrorNumber)
				i, err := strconv.Atoi(r.FindString(line))
				if err != nil {
					return nil, err
				}
				return nil, errorFromCode(i)
			}
		} else {
			// valid responses are formatted as:
			//
			//      To: <number> ID: <id>
			//
			extractTo := regexp.MustCompile(matchTo)
			extractID := regexp.MustCompile(matchID)

			to := strings.Split(extractTo.FindString(line), " ")[1]
			id := strings.Split(extractID.FindString(line), " ")[1]

			nums[to] = map[string]string{"ID": id}
		}
	}

	if badNumbers {
		if len(nums) > 0 {
			return nums, ErrInvalidTo
		}
		return nil, ErrInvalidTo
	}

	return nums, nil
}

// parseCreditResponseBody parse the plain text response body from a clockwork
// /credit HTTP call.
func parseCreditResponseBody(body string) (float64, string, error) {
	if strings.Contains(body, "Error") {
		// error response is in the following plain-text format:
		//
		//	Error 58: Invalid API Key
		//
		r := regexp.MustCompile(matchErrorNumber)
		n := r.FindString(body)
		i, err := strconv.Atoi(n)
		if err != nil {
			return 0, "", err
		}
		return 0, "", errorFromCode(i)
	}
	// valid response body is in the following plain-text format:
	//
	//	Balance: 287.58 (GBP)
	//
	// extract amount 287.58 and currency code GBP
	extractAmount := regexp.MustCompile(matchCurrency)
	extractCurrencyCode := regexp.MustCompile(matchCurrencyCode)
	a := extractAmount.FindString(body)
	c := extractCurrencyCode.FindString(body)
	s, err := strconv.ParseFloat(a, 64)
	if err != nil {
		return 0, "", err
	}
	return s, c, nil
}

// urlEncode encodes key/value pairs in a map as a HTTP Get query string. e.g.
// {"Message": "Hello World"}  -> "Message=Hello+World"
func urlEncode(queryValues map[string]string) string {
	u := url.Values{}
	for k, v := range queryValues {
		u.Add(k, v)
	}
	return u.Encode()
}

// smsSetOptions returns all the user set fields on the SMS type.
func smsSetOptions(sms SMS) map[string]string {
	vals := make(map[string]string)
	if sms.To != nil {
		vals["To"] = strings.Join(sms.To, ",")
	}
	if sms.Content != "" {
		vals["Content"] = sms.Content
	}
	if sms.From != "" {
		vals["From"] = sms.From
	}
	if sms.MsgType != "" {
		vals["MsgType"] = sms.MsgType
	}
	if sms.Concat != 0 {
		vals["Concat"] = strconv.Itoa(sms.Concat)
	}
	if sms.ClientID != "" {
		vals["ClientID"] = sms.ClientID
	}
	if sms.Expiry.Minutes() > 10 { // 10 = minimum expiry time
		vals["ExpiryTime"] = strconv.Itoa(int(sms.Expiry.Minutes()))
	}
	if !sms.AbsExpiry.IsZero() {
		vals["AbsExpiry"] = formatTime(sms.AbsExpiry)
	}
	if sms.UniqueIDChecks {
		vals["UniqueId"] = "1"
	}
	if sms.InvalidCharAction != 0 {
		vals["InvalidCharAction"] = strconv.Itoa(sms.InvalidCharAction)
	}
	if sms.Truncate != 0 {
		vals["Truncate"] = strconv.Itoa(sms.Truncate - 1)
	}
	return vals
}

// formatTime formats a time instance in the form yyyyMMddHHmm e.g. 201110201530
func formatTime(t time.Time) string {
	return fmt.Sprintf("%04d%02d%02d%02d%02d", t.Year(), t.Month(),
		t.Day(), t.Hour(), t.Minute())
}
